Utils/AsyncReducer.swift

```diff
@@ -0,0 +1,144 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Deferred
+
+private let DefaultDispatchQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)
+
+public func asyncReducer<T, U>(initialValue: T, combine: (T, U) -> Deferred<Maybe<T>>) -> AsyncReducer<T, U> {
+    return AsyncReducer(initialValue: initialValue, combine: combine)
+}
+
+/**
+ * A appendable, async `reduce`.
+ * The reducer starts empty. New items need to be `append`ed.
+ * The constructor takes an `initialValue`, a `dispatch_queue_t`, and a `combine` function.
+ * The reduced value can be accessed via the `reducer.terminal` Deferred<Maybe<T>> which is runs once all items have been combined.
+ * The terminal will not be filled if no items have been appended.
+ * Once the terminal has been filled, no more items can be appeneded, and `append` methods will error.
+ */
+public class AsyncReducer<T, U> {
+
+    public typealias Combine = (T, U) -> Deferred<Maybe<T>>
+
+    private let initialValueDeferred: Deferred<Maybe<T>>
+
+    private let dispatchQueue: dispatch_queue_t
+
+    public let terminal: Deferred<Maybe<T>> = Deferred()
+
+    private var queuedItems: [U] = []
+
+    private var index = 0
```

Comment:
`index` is unnecessary if you pop items off the front of `queuedItems` as they run — that array becomes `remaining`.